Regex fix and test

comagnaw opened the pull request:
@Grokzen I worked on this pull request from the master branch of pykwalify. 

- removed the encode method called against dictionary which was causing regex failed matches to fault.
- added 3 failing unit tests to verify regex matching-rules for default 'any', declared 'any', and declared 'all' would properly run.  
-  'all' was not an allowed unit test, so I updated the test_rule.py unit test to allow 'all'.  
-  updated validation rules doc to clarify a bit what 'any' does, as I assume it should be 'one or more'.
- updated failure messages for any or all so they were joined with 'and' or 'or'.  I believe this improves messages.  The previous join used three spaces which did not have provide enough clarity on the break-down of the mulitple regex's matched.
- updated formatting on some comments.

I do need clarity on message I found in rule.py

```
# Verify that the provided rule is part of one of the allowed one
        allowed = ["any", "all"]
# ["none", "one", "all"] Is currently awaiting proper implementation
```
I presume this an artifact from development phase as 'any' and 'all' are in the documentation as supported.

Grokzen commented:
Also you should add a note in the release notes file with a short description about your changes. Put it in a new section named `Next release` I will handle the merge conflict between this PR and the other one that is currently open. You can also (if you want) add yourself to Authors list in docs/Authors file.

Grokzen commented:
The problem you are seeing is because of a minor quirk in py3.4 where items is not read and stored in the same way as the file specifies. It is somewhat random the order that it reads and itterates over the standard `dict` data structure and that causes problems in this case with the regex keys being read from file randomly.

Grokzen commented:
The only good way currently is limit the test to check for the string `Key 'foobar3' does not match all regex` instead of the entire string and then possibly check if `'^foobar.*$'` and `'^.*[1-2]$'`` is inside the string but you do not really care in what order they are present.

The longterm solution would be to change how json and yaml is read to only work with OrderedDict:s that will order in a way that can be tested.

Grokzen commented:
I merged the other PR so you should be able to rebase your code ontop of the new commits and resolv the conflicts on your end.

comagnaw commented on pykwalify/core.py:
```diff
@@ -486,16 +486,17 @@ def _validate_mapping(self, value, rule, path, errors, done=None):
+                            value=value,
                             key=k,
-                            regex="  ".join([mm[0]._map_regex_rule for mm in regex_mappings])))
+                            regex="' or '".join(sorted([mm[0]._map_regex_rule for mm in regex_mappings]))))
```
@Grokzen I believe sorted method on the list comprehension has made the error message more predictable.   Local tests and CI reports good results.

Grokzen replied:
Nice find. That will make it more reliable and testable.

Grokzen commented on pykwalify/rule.py:
```diff
@@ -182,7 +182,7 @@ def init_matching_rule(self, v, rule, path):
         # Verify that the provided rule is part of one of the allowed one
-        allowed = ["any"]
+        allowed = ["any", "all"]
         # ["none", "one", "all"] Is currently awaiting proper implementation
```
Remove `all` from the commented out line.

Grokzen commented on ReleaseNotes.rst:
```diff
@@ -7,6 +7,7 @@ Next Release
 
 - float / number type now support range restrictions
 - ranges on non number types (e.g. seq, string) now need to be non negative.
+- Fixed regex issue #22 and added regex unit tests.
```
Please try to make this line more descriptive what problem was fixed and there is no need to link to the issue in the release notes.

comagnaw commented:
@Grokzen I went ahead and added commit to fix issue #25 and added unit tests to cover.  Release notes are updated to reflect the additional change.  Let me know if there is anything else needed for this PR.

Grokzen commented:
@comagnaw I will review it in detail later today or possibly tomorrow and get back to you

Grokzen commented:
@comagnaw LGTM, merging and will cut a release either later tonight or tomorrow
